datastream-kafka/src/test/java/com/linkedin/datastream/kafka/TestKafkaProducerWrapper.java

```diff
@@ -0,0 +1,152 @@
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import com.codahale.metrics.MetricRegistry;
+
+import com.linkedin.datastream.common.Datastream;
+import com.linkedin.datastream.metrics.DynamicMetricsManager;
+import com.linkedin.datastream.server.DatastreamTask;
+import com.linkedin.datastream.server.DatastreamTaskImpl;
+import com.linkedin.datastream.testutil.DatastreamTestUtils;
+
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.anyLong;
+import static org.mockito.Mockito.doThrow;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+
+/**
+ * Tests for {@link KafkaProducerWrapper}
+ */
+@Test
+public class TestKafkaProducerWrapper {
+
+  @Test
+  public void testFlushInterrupt() throws Exception {
+    DynamicMetricsManager.createInstance(new MetricRegistry(), getClass().getSimpleName());
+    Properties transportProviderProperties = new Properties();
+    transportProviderProperties.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:1234");
+    transportProviderProperties.put(ProducerConfig.CLIENT_ID_CONFIG, "testClient");
+    transportProviderProperties.put(KafkaTransportProviderAdmin.ZK_CONNECT_STRING_CONFIG, "zk-connect-string");
+
+    String topicName = "random-topic-42";
+
+    MockKafkaProducerWrapper<byte[], byte[]> producerWrapper =
+        new MockKafkaProducerWrapper<>("log-suffix", transportProviderProperties, "metrics");
+
+    String destinationUri = "localhost:1234/" + topicName;
+    Datastream ds = DatastreamTestUtils.createDatastream("test", "ds1", "source", destinationUri, 1);
+
+    DatastreamTask task = new DatastreamTaskImpl(Collections.singletonList(ds));
+    ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(topicName, null, null);
+    producerWrapper.assignTask(task);
+
+    // Sending first event, send should pass, none of the other methods on the producer should have been called
+    producerWrapper.send(task, producerRecord, null);
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(0);
+    producerWrapper.verifyClose(0);
+    Assert.assertEquals(producerWrapper.getNumCreateKafkaProducerCalls(), 1);
+
+    // Calling the first flush() on a separate thread because the InterruptException calls Thread interrupt() on the
+    // currently running thread. If not run on a separate thread, the test thread itself will be interrupted.
+    ExecutorService executorService = Executors.newSingleThreadExecutor();
+    executorService.submit(() -> {
+      // Flush has been mocked to throw an InterruptException
+      Assert.assertThrows(InterruptException.class, producerWrapper::flush);
+    }).get();
+
+    producerWrapper.verifySend(1);
+    producerWrapper.verifyFlush(1);
+    producerWrapper.verifyClose(1);
+
+    // Second send should create a new producer, resetting flush() and close() invocation counts
```

Comment:
nit: resetting send, flush and close (for completeness)

Reply:
As discussed offline, I'll take care of this on the next review. :)